src/NLog/GlobalDiagnosticsContext.cs

```diff
@@ -65,13 +62,29 @@ public static void Set(string item, string value)
         /// <param name="item">Item name.</param>
         /// <param name="value">Item value.</param>
         public static void Set(string item, object value)
+        {
+            SetInternal(item, value);
+        }
+
+        private static void SetInternal(string item, object value)
         {
             lock (_dict)
             {
-                GetWritableDict(_dictReadOnly != null && !_dict.ContainsKey(item))[item] = value;
+                if (ContainsItemInternal(item))
+                {
+                    var newDict = CloneDictionary();
+                    SetWritableDictionary(newDict);
+                }
+
+                _dict[item] = value;
             }
         }
 
+        private static bool ContainsItemInternal(string item)
+        {
+            return _dictReadOnly != null && !_dict.ContainsKey(item);
```

Comment:
ContainsKey with not-operator doesn't match name `ContainsItemInternal`

Reply:
fixed